[소니] 로또 1단계 - 자동 미션 제출합니다.

sonypark opened the pull request:
화투님 안녕하세요. 우아한테크코스 2기 소니 입니다.
이번 미션은 아직 입출력 부분을 구현중이라 구현한 부분까지만 PR을 보냅니다.
도메인 부분은 모두 구현 완료했습니다.
입출력 부분은 오늘 자정 전까지 구현해서 추가 푸시하겠습니다.
감사합니다!

sonypark committed:
docs: 미션개요, 구현상세, 기능목록 작성

sonypark committed:
critical: 도메인, 뷰 설계

sonypark committed:
docs: 클래스명 변경으로 인한 문서 수정: 발급 -> 생성

sonypark committed:
feat: 임의로 1부터 45까지의 숫자를 6개 생성하여 리스트로 만든다

sonypark committed:
refac: 임의로 1부터 45까지의 숫자를 6개 생성하여 리스트로 만든다

sonypark committed:
docs: '로또 번호 중복 허용 불가' 기능 추가

sonypark committed:
feat: 로또 숫자 생성 시 6개의 숫자는 중복되지 않는다

sonypark committed:
refac: 테스트 메서드명 변경: generate -> 로또숫자_생성_갯수

sonypark committed:
feat: 구매 금액을 생성할 로또 개수로 변환한다

sonypark committed:
feat: 로또 개수만큼 로또를 생성한다

sonypark committed:
refac: 로또 개수만큼 로또를 생성한다

sonypark committed:
feat: 생성된 로또와 지난 주 당첨번호를 비교한다

sonypark committed:
refac: LottosGenerator의 generate() 인자값 변경

Lotto 생성자 변경에 따라 LottosGenerator의 generate() 인자값을 변경함

sonypark committed:
refac: Lotto의 match() 함수를 matchWithWinningNumbers()로 변경

이전에 bonusBall을 한번에 비교하려고 잘못 생각해서 변경

sonypark committed:
feat: 로또번호와 보너스볼 비교(일치 여부 확인)

sonypark committed:
docs: 당첨 금액 규칙 중 명칭 변경

기존에 일치한 숫자로 표현한 것을 등수로 변경

sonypark committed:
refac: LottoNumbersGenerator의 generate() 로직 변경

기존에 로직 반영이 잘 안된 부분을 수정

sonypark committed:
feat: 당첨 금액 산출 규칙에 따라 순위를 산출한다.

sonypark committed:
feat: 당첨 금액 산출 규칙에 따라 당첨 금액을 산출한다

sonypark committed:
feat: 수익률을 계산한다

sonypark committed:
feat: 총 당첨 금액을 계산한다

sonypark committed:
feat: 로또 게임을 구동한다

sonypark committed:
ing: 출력 부분 진행중입니다.

sonypark committed:
chore: 제출 전 급하게 구현한 출력 로직 삭제

sonypark committed:
feat: 수익률을 출력한다

sonypark committed:
feat: 결과를 담는 객체를 만들고, 결과를 추가

sonypark committed:
feat: 랭크별 당첨된 티켓 수 조회

sonypark committed:
feat: 당첨 통계 출력

sonypark committed:
refac: 당첨 결과를 계산

sonypark committed:
refac: 로또 게임 구동

sonypark commented:
입출력 로직을 구현해서 추가 커밋 완료했습니다.
아직 프로그래밍 요구사항을 반영하지 못한 부분이 있습니다.
- indent(인덴트, 들여쓰기) depth를 2단계에서 1단계로 줄여라.
- 메소드의 크기가 최대 10라인을 넘지 않도록 구현한다.
- 모든 원시값과 문자열을 포장한다.
- 일급 콜렉션을 쓴다.

금요일까지 프로그래밍 요구사항대로 리팩토링 한 후 추가 푸시하겠습니다.
감사합니다!

sonypark committed:
test: StringUtils 테스트 추가

sonypark committed:
refac: StringUtils 안의 변수명 변경(STANDARD -> DELIMITER)

sonypark committed:
refac: LottoNumbersGenerator 내 함수 분리

이외에 random에 대한 학습 테스트, 접근 제한자 수정 등의 작업이 있음

sonypark committed:
refac: Rank의 함수명 변경(isExists -> isValid)

sonypark committed:
refac: lottoNumbers의 타입 변환(List -> Set)

이외에 convertMoneyToLottosSize() 함수의 위치 변경, lottoNumbersString의
네이밍 변경

sonypark committed:
refac: EarningRateCalculator 상수명 변경

sonypark commented:
현재 - 금요일 18:30분 - 리팩토링 구현한 부분까지 완료했습니다.

아직 구현하지 못한 부분
- LottoGame, Result 클래스 부분에 대한 리팩토링은 (진행 중)
- 전체 코드 예외 처리

현재 페어와 같이 조금 느리더라도 끝까지 함께 페어 프로그래밍을 유지하면서 코드를 완성해나가고 있습니다.
지금까지 구현된 부분에 대해 피드백 주면 감사하겠습니다.
(특별히 궁금한 부분에 대해서는 주석으로 달아 놓았습니다.)

phs1116 commented on src/main/java/lotto/domain/EarningRateCalculator.java:
```diff
@@ -0,0 +1,13 @@
+public class EarningRateCalculator {
+
+    /* PERCENTILE은 백분위를 의미합니다 */
+    private static final int PERCENTILE = 100;
```
PERCENTILE는 이럴 때 쓰는 용어가 아니지 않나요? :)

sonypark replied:
네 말씀하신대로 PERCENTILE용어는 적절치 않은 것 같아. PERCENT로 수정하였습니다.

phs1116 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,34 @@
+    }
+    int matchWinningNumbers(Set<Integer> winningNumbers) {
+        // 질문: List구현체가 ArrayList일 경우, parallel을 쓰면 속도가 빨라진다고 해서 써 봤는데요. 혹시 잘못 쓴 것은 아닌지, parallel을 언제 쓰면 좋은 지 궁금합니다.
+        return (int)lottoNumbers.stream()
```
parallel자체가 멀티 쓰레드로 여러 쓰레드에서 동작하는것이기 때문에 그저 '빠르다'라는 이유로 사용하기에는 위험성이 커요!
대량 데이터에 대해서나 의미가 있지 크지 않는 데이터에 대해서는 쓰레드 사용 비용에 효과는 미미하구요.
사용하시려면 쓰레드풀까지 고민하시면서 사용하는게 맞지 현재 단계에서는 굳이 사용하지 않으셔도 됩니다.
실제로 특별 케이스가 아닌 이상 많이 쓰이지도 않아요! :)

sonypark replied:
아 그렇군요! 사실 parallel을 쓰는 코드를 아직 보지 못했는데 어떤 경우 사용하는게 좋은지 싶어 질문드렸습니다. 답변 감사합니다!

phs1116 commented on src/main/java/lotto/domain/LottoComparator.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.List;
+import java.util.Set;
+
+class LottoComparator {
```
로또 자체에 이미 해당 기능이 있는데 클래스로 분리할 필요가 있을까요? :)

sonypark replied:
네 해당 클래스는 삭제하고 Lotto 클래스에서 처리하도록 했습니다.

phs1116 commented on src/main/java/lotto/domain/LottoConverter.java:
```diff
@@ -0,0 +1,8 @@
+package lotto.domain;
+
+public class LottoConverter {
```
Money라는 값객체를 활용해보면 어떨까요?

sonypark replied:
금액 관련된 부분은 Money 객체로 감싸주었습니다.

phs1116 commented on src/main/java/lotto/domain/LottoNumberConfig.java:
```diff
@@ -0,0 +1,7 @@
+    static final int MAX = 45;
+    static final int MIN = 1;
+    static final int SIZE = 6;
+}
```
응집력 있게 해당 값이 필요한 객체가 들고있는게 맞을 것 같아요.
이렇게 전역 변수의 형태로 값을 꺼내놓으면 어디서 어떻게 영향을 줄 지에 대해 파악도 쉽지 않아요.

sonypark replied:
네 해당 상수가 로또 번호에만 쓰이는 거라 로또 번호 클래스로 이동했습니다.

phs1116 commented on src/main/java/lotto/domain/LottoNumbersGenerator.java:
```diff
@@ -0,0 +1,43 @@
+     * 1. List<Integer>로 둘 경우, 메모리 문제나 성능 상 문제가 생기지 않을까 걱정이 됐습니다.
+     * 2. 가독성 측면에서 void가 나은 지, List<Integer>가 나은 지 헷갈렸습니다.
+     * 이에 대해 가르침 주시면 감사하겠습니다.
+     */
```
새로운 객체가 생성되는 것이 아니기 때문에 성능과 메모리 상에 문제는 없어요. :)
자바의 메모리 구조에 대해서 공부해보시면 좋을 것 같아요.

sonypark replied:
네 조언 감사합니다. 자바 메모리 구조에 대해 공부해보겠습니다!

phs1116 commented on src/main/java/lotto/domain/LottoNumbersGenerator.java:
```diff
@@ -0,0 +1,43 @@
+
+    private static int pickLottoNumber() {
+        return random.nextInt(LottoNumberConfig.MAX - LottoNumberConfig.MIN) + LottoNumberConfig.MIN;
+    }
```
어차피 1-45번이라는 숫자는 정해져 있으니 매번 랜덤으로 생성하기보다는
1-45개 숫자만큼의 로또를 만들어두고 셔플링 하는 방식이 더 간결하고 성능도 낫지 않을까요?
이러면 중복 체크와 6개 채우려고 반복할 필요도 없구요. :)

sonypark replied:
맞습니다. 매번 랜덤 넘버를 생성할 필요없이 미리 만들어주고 셔플하는게 훨씬 효율적이네요. 말씀하신대로 수정하였습니다!

phs1116 commented on src/main/java/lotto/domain/LottosGenerator.java:
```diff
@@ -0,0 +1,20 @@
+public class LottosGenerator {
+
+    static List<Lotto> generate(int lottosSize, List<Set<Integer>> lottoNumbersList) {
+        //질문: final을 붙이는 것이 의미가 있는 것인 지 궁금합니다. 제 생각에는 이후에 초기화를 더 하지 못하게 막음으로써 메모리 누수를 막을 수 있다고 생각하는데요. 어떻게 생각하시나요?
```
의미는 있어요. 메모리 누수가 우려될정도로 성능적인 문제는 아니에요.
자바 어플리케이션에서는 그정도로 성능 고민 하지 않으셔도 충분히 JVM과 서버 성능이 뛰어나서요. :)
가장 큰 장점은 불변 보장으로 사이드이팩트 방지가 크다고 보시면 될 것 같아요.

sonypark replied:
`final`을 사용하는 목적은 성능적인 문제보다 사이드이팩트 방지이다. 잘 기억하겠습니다 :)

phs1116 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,49 @@
+
+            if (rank.matchNumber == matchNumber) {
+                return rank;
+            }
```
if문 보다는 Rank의 enum값 자체가 스스로 자신에 해당하는지 판단하는게 객체지향 관점에서 더 알맞지 않을까요?

sonypark replied:
해당 부분을 Rank의 메서드로 만들어서 수정해보았습니다.

phs1116 commented on src/main/java/lotto/domain/TotalWinningMoneyCalculator.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.List;
+
+public class TotalWinningMoneyCalculator {
```
별도 클래스로 분리할정도의 책임긴가? 싶네요. :)

sonypark replied:
네 분리할만큼 의미있는 클래스는 아닌것 같습니다. WinningRanks 클래스에 해당 로직을 이동했습니다.

phs1116 commented on src/test/java/lotto/domain/LottoNumbersGeneratorTest.java:
```diff
@@ -0,0 +1,29 @@
+class LottoNumbersGeneratorTest {
+
+    @Test
+    void 로또숫자_생성_갯수() {
```
갯수 말고 어떤 값이 돌아오는지에 대한 테스트도 작성해보도록 노력해주세요. :)

phs1116 reviewed (changes requested):
안녕하세요, 소니! 리뷰어 화투입니다.
아직 다 구현을 하지 못하신 것 같아서 최대한 드릴수 있는데까지 피드백 추가했어요.
참고 부탁드리고 궁금하신 점 있으시면 연락주세요. :)

sonypark committed:
refac: 피드백 반영

1. 변수명 변경: PERCENTILE -> PERCENT
2. parallel 삭제
3. 클래스 변경 및 삭제: LottoComparator,LottoNumberConfig,TotalWinningMoneyCalculator 삭제 및 로직 이동
4. LottoNumbersGeneratorTest: 로또 숫자 유효범위 테스트 추가

sonypark committed:
refac: 피드백 반영 : Random객체 사용 -> Collections.shuffle() 사용

sonypark committed:
refac: 피드백 반영 - Rank의 enum값 자체가 스스로 자신에 해당하는지 판단하도록 함

sonypark committed:
refac: 원시값 포장 - Money

sonypark committed:
refac: 원시값 포장 - LottoNumber

sonypark committed:
refac: 일급 콜렉션 적용: LottoNumberBox

sonypark committed:
refac: LottoNumbersGenerator -> LottosGenerator 로 변경

sonypark committed:
refac: 당첨금액 총합과 수익률을 가지는 result 객체를 만듦

sonypark committed:
refac: 로또 게임 구동

sonypark committed:
docs: README에 사용 방법 추가

sonypark committed:
[chore] 의미없는 주석 및 사용하지 않는 import 문 삭제

ttkmw committed:
refac: 피드백 반영

sonypark committed:
[chore] 코드 라인 정리 및 메서드 위치 조정

phs1116 commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,61 @@
+
+    public void play() {
+        Money purchaseAmount = inputPurchaseAmount();
+        List<Lotto> lottos = purchaseLottos(purchaseAmount);
```
일급 컬렉션을 공부해보고 이를 적용해보면 어떨까요? :)
[일급 컬렉션](https://jojoldu.tistory.com/412)

sonypark replied:
Lottos라는 일급컬렉션을 만들어 적용해보았는데 의도하신대로 잘 한 걸까요?

phs1116 commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,61 @@
+            int matchingNumber = lotto.matchWinningNumbers(winningLotto);
+            if (Rank.isValid(matchingNumber)) {
+                Rank rank = Rank.valueOf(matchingNumber, lotto.matchBonusNumber(bonusNumber));
+                winningRanks.addWinningRanks(rank);
```
addWinningRanks이 아닌 생성자를 통해 결과값을 생성하고 WinningRanks의 생성자를 통해 생성하는게 사이드 이펙트의 여지를 많이 줄일 수 있지 않을까요? :)

`이펙티브 자바 아이템 17. 변경 가능성을 최소화하라`

sonypark replied:
이 부분은 WinningRanksFactory를 만들어 로직을 수행한 후 결과를 반환하도록 수정했습니다.

phs1116 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,40 @@
+    private static final int MIN = 1;
+    private int value;
+
+    public LottoNumber(int value) {
```
LottoNumberBox에서 캐시를 통해 값을 가지고 있음에도 다시 재생성 될 위험이 있어보여요.
생성자에 제한을 두어 재생성을 방지해보면 어떨까요?

sonypark replied:
생성자에 제한을 두라는 말씀이 생성자의 접근제어자를 private으로 하라는 말로 이해했습니다. 그럴 경우 해당 생성자를 사용하는 테스트 케이스를 테스트하기 어려워져 고민이 됩니다. 찾아보니 mocking과 같은 방법으로 할 수 있을 것 같은데 아직 그 부분에 대해서는 잘 알지 못해서 일단 보류해두었습니다.

phs1116 commented on src/main/java/lotto/domain/LottoNumberBox.java:
```diff
@@ -0,0 +1,19 @@
+
+class LottoNumberBox {
+    private static final int LOTTO_MAX_NUMBER = 45;
+    private static final int LOTTO_MIN_NUMBER = 1;
```
LottoNumber와 같은 정책을 가지고 있네요. 
LottoNumber에서 직접 캐시를 관리하도록 하여 정책과  LottoNumber 대해서 한곳에서 관리하도록 해보면 어떨까요?
이 과정에서 `이펙티브 자바 아이템 1. 생성자 대신 정적 팩터리 메서드를 고려하라`도 같이 고려해보면 좋을 것 같아요. :)

sonypark replied:
LottoNumber에서만 정책을 갖고 있고 LottoNumberBox 1~45번까지의 로또 번호를 관리하도록 했습니다.

phs1116 commented on src/main/java/lotto/domain/LottosGenerator.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+import java.util.Set;
+
+public class LottosGenerator {
```
해당 클래스는 shuffle때문에 값의 리턴에 대해서는 테스트가 어려워 보여요.
테스트하기 어려운 부분에 대해서는 인터페이스로 분리하고 원하는 값이 리턴되는지 또한 테스트해보면 어떨까요?

sonypark replied:
이 부분도 인터페이스에 대한 이해가 부족하여 공부 중 입니다. 인터페이스를 사용하긴 했는데 어떤 식으로 테스트를 진행해야할지는 감이 잘 안 잡혀서요.

phs1116 replied:
현재 잘 구현하신 것 같습니다. :)
테스트 코드상 원하는 값을 돌려주는 구현체를 주입해주시면 테스트가 용이하실겁니다.

phs1116 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,59 @@
+    private final int value;
+
+    Money() {
+        this.value = 0;
```
0같은 값은 상수로 사용해도 되지 않을까요? :)

sonypark replied:
네 상수로 뺐습니다 :)

phs1116 commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private WinningRanks compareWithWinningNumbers(List<Lotto> lottos, Lotto winningLotto, LottoNumber bonusNumber) {
+        WinningRanks winningRanks = new WinningRanks(new HashMap<>());
```
WinningRanks의 생성 주체를 절차적으로 풀지 말고 객체의 상호작용을 통해 생성하도록 해보면 어떨까요?

ex) winninLotto.match(lottos)

sonypark replied:
WinningLotto 클래스에서 당첨번호를 매칭하는 로직을 구현하여 winninLotto.match(lottos)형태로 변경했습니다.

phs1116 commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,61 @@
+
+    private Result produceResult(List<Lotto> lottos, Money purchaseMoney) {
+        Lotto winningLotto = InputView.inputLastWeekWinningNumbers();
+        LottoNumber bonusNumber = InputView.inputBonusNumber();
```
winningLotto와 bonusNumber는 결국엔 하나의 당첨번호가 아닐까요?
당첨 로또 번호라는 하나의 객체로 묶을 수 있지 않을까요?

sonypark replied:
WinningLotto라는 클래스를 만들어 winningLotto와 bonusNumber를 하나의 객체로 묶었습니다.

phs1116 reviewed (changes requested):
안녕하세요, 소니! 리뷰어 화투입니다.
구현 매우 잘해주셨어요. :)
다만 몇가지 더 개선할 수 있는 점이 있어 피드백 추가했습니다.
참고 부탁드리고 궁금하신 점 있으시면 연락주세요!

sonypark committed:
refac: 로또 번호의 최소,최대값을 LottoNumber에서 캐시

sonypark committed:
refac: 매직넘버를 자바 상수로 변환

sonypark committed:
feat: 당첨 로또를 만들고, 로직(match) 이동

sonypark committed:
refac: WinningRanks를 팩토리를 통해 생성

sonypark committed:
refac: Lottos 일급 콜렉션 적용

sonypark committed:
refac: Collections.shuffle()하는 부분을 인터페이스로 바꿈

sonypark committed:
[chore] 파일 위치 변경

phs1116 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,32 @@
+
+public class Lotto {
+    private static final int LOTTO_NUMBER_SIZE = 6;
+    Set<LottoNumber> lottoNumbers;
```
캡슐화를 위해 접근제어자 신경써주세요! :)

phs1116 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public Set<LottoNumber> getLottoNumbers() {
+        return lottoNumbers;
```
외부에 의해 내부 상태 변경을 방지하기 위해 Collections.unmodifiableSet등을 활용해보면 어떨까요? :)

sonypark replied:
`Collections.unmodifiableSet`으로 리턴해주어서 변경을 방지했습니다.

phs1116 commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,59 @@
+    private WinningRanks compareWithWinningNumbers(Lottos lottos, WinningLotto winningLotto) {
+        Ranks ranks = new Ranks();
+        //todo: 함수를 더 간결하게 정리
+        for (Lotto lotto : lottos.getLottos()) {
+            Rank rank = winningLotto.match(lotto);
+            ranks.add(rank);
+        }
```
`이펙티브 자바 아이템 17. 변경 가능성을 최소화하라`
add를 통한 수정보다는 리스트를 지역변수로 생성 후 생성자로 전달해보면 어떨까요?
혹은 `Ranks ranks = winningLotto.match(lottos);` 와 같이 더 추상화 할 수 있지 않을까요?

sonypark replied:
`Ranks ranks = winningLotto.match(lottos);` 로 추상화 했습니다.

phs1116 commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    private void printLottos(Lottos lottos) {
+        for (Lotto lotto : lottos.getLottos()) {
+            OutputView.printLotto(lotto);
+        }
```
printLotto내에서 로또 리스트를 출력하도록 순회 조차 책임질수 있지 않을까요?

phs1116 commented on src/main/java/lotto/domain/LottoNumberBox.java:
```diff
@@ -0,0 +1,16 @@
+class LottoNumberBox {
+    private static final int ONE = 1;
+
+    static List<LottoNumber> create() {
+        return IntStream.range(LottoNumber.MIN, LottoNumber.MAX + ONE)
+            .mapToObj(LottoNumber::new)
+            .collect(Collectors.toList());
+    }
```
캐시에 대한 로직이 있었는데 없어진 것 같네요. 이유가 있으신가요??? :)

sonypark replied:
캐시에 대한 로직을 LottosGenerator으로 옮겼습니다.
LottosGenerator에서 한 번 호출해서 사용합니다.

```java
private static List<LottoNumber> lottoNumbersInBox = LottoNumberBox.create();
```

phs1116 commented on src/main/java/lotto/domain/LottosGenerator.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    Lottos generate(int lottosSize) {
+        Set<Lotto> lottos = new HashSet<>();
+        for (int i = 0; i < lottosSize; i++) {
+            Set<LottoNumber> lottoNumbers = generateLottoNumbers();
+            lottos.add(new Lotto(lottoNumbers));
+        }
+        return new Lottos(lottos);
```
위아 마찬가지로 
`이펙티브 자바 아이템 17. 변경 가능성을 최소화하라`
add를 통한 수정보다는 리스트를 지역변수로 생성 후 생성자로 전달해보면 어떨까요?

sonypark replied:
add를 통한 수정이 아닌 생성자로 전달하는 방법은 잘 이해가 되지 않는데 혹시 추가 설명 해주실 수 있을까요?

phs1116 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,51 @@
+            return SECOND;
+        }
+
+        if (matchNumber < FIFTH.matchNumber) {
+            return null;
+        }
```
null을 전달하여 null 위험성을증가시키고 사용하는 쪽에서 null을 체크하게 하는것 보다는 당첨되지 않았다는 의미의 enum값을 넘겨주는게 더 좋지 않을까요? :)

phs1116 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static Rank valueOf(int matchNumber, boolean matchBonusBall) {
+        if (SECOND.isSameMatchNumber(matchNumber) && matchBonusBall) {
+            return SECOND;
+        }
```
이런 추가 요소에 대한 체크를 각 enum값들의 함수로 설정 할 수 있지 않을까요?

phs1116 commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.Set;
+
+public class WinningLotto extends Lotto {
```
`이펙티브 자바 아이템 18. 상속보다는 컴포지션을 사용하라`
참고해보고 수정해보면 좋을 것 같아요. :)

sonypark replied:
네 말씀하신대로 상속이 아닌 컴포지션을 이용하여 구현해보았습니다.

그런데 상속과 컴포지션을 언제 써야할지 공부를 하던 중 질문이 생겼습니다.
보통 상속과 컴포지션을 선택하는 기준은 다음과 같이 `IS-A` 또는 `HAS-A` 관계로 한다고 나와있는데요. (`IS-A`관계일 경우 상속, `HAS-A`관계일 경우 컴포지션) 

- A computer `HAS-A` CPU.
- A circle `IS-A` shape.
- A circle `HAS-A` point.
- A laptop `IS-A` computer.
- A vector `IS-A` list.

`WinningLotto`는 `Lotto`와 비교했을 때 `IS-A` 관계라고 여겨지는데 그러면 상속으로 구현하는게 맞지 않나 싶어서요. 이 부분에 대한 답변 주시면 감사하겠습니다!

- 참고: [Composition vs. Inheritance](https://dzone.com/articles/composition-vs-inheritance)

phs1116 commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,26 @@
+        this.bonusNumber = bonusNumber;
+    }
+
+    Rank match(Lotto lotto) {
```
접근제어자 신경써주세요!

sonypark replied:
Rank의 match 함수는 LottoGame에서 사용해서 디폴트 접근제어자로 선언했는데, 혹시 문제가 있을까요?

phs1116 reviewed (approved):
안녕하세요, 소니! 
리뷰어 화투입니다.
구현 잘해주셨어요. 👍 
이번 단계 머지할게요.
다만 피드백 몇가지 추가했는데 다음 단계 진행하시면서 참고 부탁드릴게요.
그리고 한번씩 접근제어자도 확인 부탁드려요. :)
